pkg/skaffold/build/ko/build_test.go

```diff
@@ -0,0 +1,153 @@
+)
+
+func TestBuildKoImages(t *testing.T) {
+	tests := []struct {
+		description         string
+		ref                 string
+		imageID             string
+		pushImages          bool
+		importpath          string
+		imageNameFromConfig string
+		workspace           string
+	}{
+		{
+			description:         "simple image name and sideload image",
+			ref:                 "gcr.io/project-id/test-app1:testTag",
+			imageID:             "imageID1",
+			pushImages:          false,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/ko",
+			imageNameFromConfig: "test-app1",
+		},
+		{
+			description:         "ko import path image name and sideload image",
+			ref:                 "gcr.io/project-id/example.com/myapp:myTag",
+			imageID:             "imageID2",
+			pushImages:          false,
+			importpath:          "ko://example.com/myapp",
+			imageNameFromConfig: "ko://example.com/myapp",
+		},
+		{
+			description:         "simple image name and push image",
+			ref:                 "gcr.io/project-id/test-app2:testTag",
+			imageID:             "testTag",
+			pushImages:          true,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/ko",
+			imageNameFromConfig: "test-app2",
+		},
+		{
+			description:         "ko import path image name and push image",
+			ref:                 "gcr.io/project-id/example.com/myapp:myTag",
+			imageID:             "myTag",
+			pushImages:          true,
+			importpath:          "ko://example.com/myapp",
+			imageNameFromConfig: "ko://example.com/myapp",
+		},
+		{
+			description:         "workspace is not cwd",
+			ref:                 "gcr.io/project-id/example.com/test-app3:myTag",
+			imageID:             "imageID3",
+			pushImages:          false,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			imageNameFromConfig: "test-app3",
+			workspace:           "../docker",
+		},
+		{
+			description:         "ko import path image name and workspace is not cwd",
+			ref:                 "gcr.io/project-id/example.com/test-app4:myTag",
+			imageID:             "imageID4",
+			pushImages:          false,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			imageNameFromConfig: "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			workspace:           "../docker",
+		},
+		{
+			description:         "ko import path image name and workspace is not cwd and import path is subdirectory of cwd",
+			ref:                 "gcr.io/project-id/example.com/test-app5:myTag",
+			imageID:             "imageID5",
+			pushImages:          false,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			imageNameFromConfig: "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			workspace:           "..",
+		},
```

Comment:
I found this test hard to puzzle through: it wasn't clear what are the inputs vs expected test results, and what are test incidentals.  `test.ref` seems an incidental?  It might help to embed the Artifact definition in the test structure?
```
{
    description: ...,
    artifact: &latestV1.Artifact{
        ArtifactType: latestV1.ArtifactType{KoArtifact: &latestV1.KoArtifact{}},
        ImageName: "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/ko",
        Workspace: ...
    },
},
```

Reply:
I agree, it's not a great test. Since it's a unit test, I didn't want to actually build or publish - that's for the integration test.

That means the test covers wiring the components together, and a few bits of functionality: determining the Go import path, ensuring the image name is printed to the `out` writer, and returning the image identifier correctly for pushed vs non-pushed images.

All conditional logic are in separate functions, so I'll replace this test with individual tests for those functions. I'll leave one test case for the `Build()` function, just to check that everything is wired together.